### PR TITLE
Add  new `ConstraintValidator` type : `MultipleEmailsWithSeparator`

### LINKED ISSUE #24616: BO - Logs - There is no email address check
#### Describe the bug

In the BO > Advanced Parameters > Logs page, in the logs by email section, there is no check of a valid address mail or not

#### Steps to Reproduce

1. Go to BO > Advanced Parameters > Logs page
2. In the logs by email section > set the field `Send emails to` to `test`
3. Save
4. See error: no validation (the type of the field should be email and not text)

This field is an improvement in 178x

**Screenshots**

#### Additional informationBO_log
* PrestaShop version: 178x
* PHP version: 7.4

### PR DESCRIPTION
**Pull request by PululuK**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add  new `ConstraintValidator` type : `MultipleEmailsWithSeparator`. 
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/24616
| How to test?      | See description below.
| Possible impacts? | N/A

#### Wrong mails in form : 
image

#### Form validation : 

image

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/24622)

### CODE REVIEW

File: src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            if (!Validate::isEmail($email)) {
+                $invalidEmails[] = $email;
+            }
+        }
+
+        if (!empty($invalidEmails)) {
+            $message = $constraint->message ?? $translator->trans(
+                'Invalid email(s) : %invalid_emails%.',
+                ['%invalid_emails%' => implode(',', $invalidEmails)],
+                'Admin.Global.Notification'
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php]:
```suggestion
                'Admin.Notifications.Error'
```

File: src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        $invalidEmails = [];
+
+        foreach ($emailsList as $email) {
+            if (!Validate::isEmail($email)) {
+                $invalidEmails[] = $email;
+            }
+        }
+
+        if (!empty($invalidEmails)) {
+            $message = $constraint->message ?? $translator->trans(
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php]: It would be better to have both a singular and a plural option. Indeed, the "(s)" is not easy to translate in some languages. We could either have two strings: 'Invalid email:' and 'Invalid emails:' or use:
transChoice('Invalid email:'|'Invalid emails:')

https://symfony.com/doc/3.4/translation.html#pluralization 

PS: There is no space before a colon in English. :-)

**Review by Julievrz — CHANGES_REQUESTED** [review]
A few changes :) Thank you in advance

File: src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php (hunk 2 above)
Comment (PululuK) [reply src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php]: @Julievrz maybe a generic wording, something like : "**Contains invalid emails**" wyf ?
Can you please suggest me a generic wording for both cases please ?

Comment (Julievrz) [reply src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php]: Hi! Your suggestion is still plural. :-) If it is not possible to provide both singular and plural options, what do you think of: 

'Test emails could not be sent to "wrongemailaddress1", "wrongemailaddress2" because the email format is invalid." 

**But** this wording would only work if the test emails were sent to the correct email addresses. Does the fact that some of the addresses entered are false cancel all the mailings (even for the correct addresses)?

File: src/Adapter/Validate.php [hunk 3]
```diff
@@ -135,4 +135,28 @@ public function isLinkRewrite($value)
...
     }
+
+    /**
+     * Check for a given email validity.
+     *
+     * @param string $email
+     *
+     * @return bool
+     */
+    public static function isEmail($email)
```
Comment (Progi1984) [inline src/Adapter/Validate.php]:
```suggestion
    public function isEmail(string $email): bool
```

File: src/Adapter/Validate.php [hunk 4]
```diff
@@ -135,4 +135,28 @@ public function isLinkRewrite($value)
...
+    }
+
+    /**
+     * Check if $data is a string.
+     *
+     * @param string $data Data to validate
+     *
+     * @return bool Validity is ok or not
+     */
+    public static function isString($data)
```
Comment (Progi1984) [inline src/Adapter/Validate.php]:
```suggestion
    public function isString(string $data): bool
```

File: tests/Unit/Adapter/ValidateTest.php [hunk 5]
```diff
@@ -51,4 +51,60 @@ public function getIsOrderWay(): iterable
...
+            [false, true],
+            [false, false],
+            [false, null],
+            [false, 666],
+            [false, -5864],
+            [false, 0],
+        ];
+    }
+
+    public function isEmailDataProvider()
```
Comment (Progi1984) [inline tests/Unit/Adapter/ValidateTest.php]:
```suggestion
    public function isEmailDataProvider(): array
```

File: tests/Unit/Adapter/ValidateTest.php [hunk 6]
```diff
@@ -51,4 +51,60 @@ public function getIsOrderWay(): iterable
...
+
+    /**
+     * @dataProvider isStringDataProvider
+     */
+    public function testIsString($expected, $data)
+    {
+        $this->assertSame($expected, Validate::isString($data));
+    }
+
+    public function isStringDataProvider()
```
Comment (Progi1984) [inline tests/Unit/Adapter/ValidateTest.php]:
```suggestion
    public function isStringDataProvider(): array
```

File: tests/Unit/Adapter/ValidateTest.php [hunk 7]
```diff
@@ -51,4 +51,60 @@ public function getIsOrderWay(): iterable
...
+     */
+    public function testIsEmail($expected, $email)
+    {
+        $this->assertSame($expected, Validate::isEmail($email));
+    }
+
+    /**
+     * @dataProvider isStringDataProvider
+     */
+    public function testIsString($expected, $data)
```
Comment (Progi1984) [inline tests/Unit/Adapter/ValidateTest.php]:
```suggestion
    public function testIsString(bool $expected, $data): void
```

File: tests/Unit/Adapter/ValidateTest.php [hunk 8]
```diff
@@ -51,4 +51,60 @@ public function getIsOrderWay(): iterable
         yield [1, 'asc'];
         yield [1, 'desc'];
     }
+
+    /**
+     * @dataProvider isEmailDataProvider
+     */
+    public function testIsEmail($expected, $email)
```
Comment (Progi1984) [inline tests/Unit/Adapter/ValidateTest.php]:
```suggestion
    public function testIsEmail(bool $expected, $email): void
```

File: tests/Unit/PrestaShopBundle/Form/Validator/MultipleEmailsWithSeparatorValidatorTest.php [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+use PrestaShopBundle\Form\Validator\Constraints\MultipleEmailsWithSeparator;
+use PrestaShopBundle\Form\Validator\Constraints\MultipleEmailsWithSeparatorValidator;
+use stdClass;
+use Symfony\Component\Validator\Test\ConstraintValidatorTestCase;
+
+class MultipleEmailsWithSeparatorValidatorTest extends ConstraintValidatorTestCase
+{
+    public const TEST_EMAILS_SEPARATOR = '|';
+
+    protected function createValidator()
```
Comment (Progi1984) [inline tests/Unit/PrestaShopBundle/Form/Validator/MultipleEmailsWithSeparatorValidatorTest.php]:
```suggestion
    protected function createValidator(): MultipleEmailsWithSeparatorValidator
```

File: src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+{
+    public function validate($value, Constraint $constraint)
+    {
+        if (!$constraint instanceof MultipleEmailsWithSeparator) {
+            throw new UnexpectedTypeException($constraint, MultipleEmailsWithSeparator::class);
+        }
+
+        $translator = (new LegacyContext())->getContext()->getTranslator();
+
+        if (!Validate::isString($value)) {
```
Comment (Progi1984) [inline src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php]: Can't you call the validator in the constructor ? Better than use sttic call

File: src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+            throw new \InvalidArgumentException('Value must be string. Input was: ' . \gettype($value));
```
Comment (Progi1984) [inline src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php]: Move to use sections

**Review by Progi1984 — COMMENTED** [review]
Some comments

File: src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php (hunk 10 above)
Comment (PululuK) [reply src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php]: yep

File: src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php (hunk 2 above)
Comment (PululuK) [reply src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php]: Done @Julievrz ! Thanks

File: src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        foreach ($emailsList as $email) {
+            if (!$this->validateAdapter->isEmail($email)) {
+                $invalidEmails[] = $email;
+            }
+        }
+
+        if (!empty($invalidEmails)) {
+            $nbInvalidEmails = count($invalidEmails);
+            $message = $constraint->message ?? $this->translator->transChoice(
+                'Invalid email : %invalid_emails%.|Invalid emails : %invalid_emails%.',
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php]:
```suggestion
                'Invalid email: %invalid_emails%.|Invalid emails: %invalid_emails%.',
```

**Review by Julievrz — DISMISSED** [review]

File: src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparator.php [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\Exception\MissingOptionsException;
+
+class MultipleEmailsWithSeparator extends Constraint
+{
+    public const INVALID_EMAILS_ERROR_CODE = 'INVALID_EMAILS_ERROR_CODE';
+
+    public $separator;
+    public $message;
```
Comment (Progi1984) [inline src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparator.php]: PHPDoc, please

File: tests/Unit/Adapter/ValidateTest.php [hunk 14]
```diff
@@ -51,4 +67,52 @@ public function getIsOrderWay(): iterable
...
+     */
+    public function testIsEmail(bool $expected, string $email): void
+    {
+        $this->assertSame($expected, $this->validate->isEmail($email));
+    }
+
+    /**
+     * @dataProvider isStringDataProvider
+     */
+    public function testIsString(bool $expected, $data): void
```
Comment (Progi1984) [inline tests/Unit/Adapter/ValidateTest.php:145]:
```suggestion
    public function testIsString(bool $expected, string $data): void
```

File: tests/Unit/PrestaShopBundle/Form/Validator/MultipleEmailsWithSeparatorValidatorTest.php [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+use Exception;
+use InvalidArgumentException;
+use PrestaShop\PrestaShop\Adapter\Validate;
+use PrestaShopBundle\Form\Validator\Constraints\MultipleEmailsWithSeparator;
+use PrestaShopBundle\Form\Validator\Constraints\MultipleEmailsWithSeparatorValidator;
+use stdClass;
+use Symfony\Component\Translation\TranslatorInterface;
+use Symfony\Component\Validator\Test\ConstraintValidatorTestCase;
+use TypeError;
+
```
Comment (Progi1984) [inline tests/Unit/PrestaShopBundle/Form/Validator/MultipleEmailsWithSeparatorValidatorTest.php:41]:
```suggestion
use function get_class;

```

File: tests/Unit/PrestaShopBundle/Form/Validator/MultipleEmailsWithSeparatorValidatorTest.php [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    /**
+     * @param mixed $multipleEmailsWithSeparator
+     *
+     * @dataProvider exceptionsInvalidMultipleEmailsWithSeparatorProvider
+     */
+    public function testExceptionsInvalidMultipleEmailsWithSeparator(Exception $expectedException, $multipleEmailsWithSeparator): void
+    {
+        $expectedExceptionClassName = \get_class($expectedException);
```
Comment (Progi1984) [inline tests/Unit/PrestaShopBundle/Form/Validator/MultipleEmailsWithSeparatorValidatorTest.php]:
```suggestion
        $expectedExceptionClassName = get_class($expectedException);
```

**Review by Progi1984 — COMMENTED** [review]
Small comments and it's ok

**Comment by Progi1984** [comment]
@PululuK Please rebase your PR.

**Comment by PululuK** [comment]
> @PululuK Please rebase your PR.

To 1.7.8.x ?

**Comment by Progi1984** [comment]
@PululuK Yes, you can.

File: src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+
+        $invalidEmails = [];
+
+        foreach ($emailsList as $email) {
+            if (!$this->validateAdapter->isEmail($email)) {
+                $invalidEmails[] = $email;
+            }
+        }
+
+        if (!empty($invalidEmails)) {
```
Comment (ghost) [inline src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php:76]: I don't know why but transchoice don't work for me. It's work only if i add {1}

http://www.okom3pom.com/Pr!vaTe/plural.webm

```
$message = $constraint->message ?? $this->translator->transChoice(
                '{1}Invalid email: %invalid_emails%.|Invalid emails: %invalid_emails%.',
                $nbInvalidEmails,
                ['%invalid_emails%' => implode(',', $invalidEmails)],
                'Admin.Notifications.Error'
            );
```

**Comment by Progi1984** [comment]
@PululuK I misread the issue. You can continue to target `develop` branch. (But you need to rebase).

**Comment by PululuK** [comment]
> @PululuK I misread the issue. You can continue to target `develop` branch. (But you need to rebase).

Yep ! Done

File: src/Adapter/Validate.php [hunk 18]
```diff
@@ -135,4 +135,28 @@ public function isLinkRewrite($value)
...
+    }
+
+    /**
+     * Check if $data is a string.
+     *
+     * @param string $data Data to validate
+     *
+     * @return bool Validity is ok or not
+     */
+    public function isString(string $data): bool
```
Comment (PierreRambaud) [inline src/Adapter/Validate.php]:
```suggestion
     * @param mixed $data Data to validate
     *
     * @return bool Validity is ok or not
     */
    public function isString($data): bool
```

As we want to test if it's a string, I'm not sure it's a good idea to add the type as parameter.
Otherwise the method seems useless :thinking:

**Review by Progi1984 — DISMISSED** [review]

File: src/Adapter/Validate.php (hunk 18 above)
Comment (PululuK) [reply src/Adapter/Validate.php]: Used in `MultipleEmailsWithSeparatorValidator.php`

**Review by Progi1984 — DISMISSED** [review]

File: tests/Unit/Adapter/ValidateTest.php [hunk 19]
```diff
@@ -96,4 +97,58 @@ public function isUnsignedIntProvider(): array
...
+    public function isStringDataProvider(): array
+    {
+        return [
+            [true, 'jdkiizdù%'],
+            [true, ' '],
+            [true, '666xlkQM'],
+            [true, '666'],
+            [true, 'true'],
+            [true, 'false'],
+            [true, 'null'],
```
Comment (PierreRambaud) [inline tests/Unit/Adapter/ValidateTest.php]:
```suggestion
            [false, true],
            [false, false],
            [false, null],
```

Comment (PululuK) [reply tests/Unit/Adapter/ValidateTest.php]: Yep !

**Review by PierreRambaud — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by Julievrz — DISMISSED** [review]
Wording ✔️

**Review by matthieu-rolland — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hello @PululuK,

I have an issue following these steps:
1. Go to BO > Advanced Parameters > Logs page 
2. In the "Logs by email" section, set the "Send emails to" filed to `test` (invalid address mail)
3. Click on save
4. See the error: the alert is duplicated twice

image

Thanks to check and feedback.

**Review by Progi1984 — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Comment by Julievrz** [comment]
Hi @PululuK 

I talked with @jolelievre and it seems that TransChoice is deprecated in Symphony 4.4, and will no longer work the same way with Symphony 5. So, unfortunately, it's risky to use TransChoice because it might not work in future versions. 

I suggest creating two strings: one singular, one plural, and display the right one depending on the number of emails, if that's possible.

Thank you for your patience as we were discussing the issue.

File: src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php [hunk 20]
```diff
@@ -89,4 +83,21 @@ public function validate($value, Constraint $constraint)
             ;
         }
     }
+
+    protected function getInvalidMailsMessage(array $invalidEmails): string
+    {
+        if (count($invalidEmails) > 1) {
+            return $this->translator->trans(
+                'Invalid emails : %invalid_emails%.',
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php]:
```suggestion
                'Invalid emails: %invalid_emails%.',
```
One last thing because there's no space before colons in English. Thank you :-)

File: src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php [hunk 21]
```diff
@@ -89,4 +83,21 @@ public function validate($value, Constraint $constraint)
...
+                ['%invalid_emails%' => implode(',', $invalidEmails)],
+                'Admin.Global.Notification'
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php]:
```suggestion
                'Admin.Notifications.Error'
```

File: src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php [hunk 22]
```diff
@@ -89,4 +83,21 @@ public function validate($value, Constraint $constraint)
...
+            );
+        }
+
+        return $this->translator->trans(
+            'Invalid email : %invalid_email%.',
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php]:
```suggestion
            'Invalid email: %invalid_email%.',
```

File: src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php [hunk 23]
```diff
@@ -89,4 +83,21 @@ public function validate($value, Constraint $constraint)
...
+            ['%invalid_email%' => implode(',', $invalidEmails)],
+            'Admin.Global.Notification'
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Validator/Constraints/MultipleEmailsWithSeparatorValidator.php]:
```suggestion
            'Admin.Notifications.Error'
```

**Review by Julievrz — DISMISSED** [review]
Wording ✔️ 
Thank you!

**Review by PierreRambaud — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @PululuK,

I have an exception when I set the email filed with space ` `. 
image

I attached a screen record
https://drive.google.com/file/d/1sAyZM6vFosbiMHW_eXzFwgnfKl8UKe9y/view?usp=sharing

Thank you!

**Review by Julievrz — APPROVED** [review]

File: src/PrestaShopBundle/Resources/config/services/bundle/form/form_validator.yml [hunk 24]
```diff
@@ -9,3 +9,11 @@ services:
       - "@translator"
     tags:
       - { name: validator.constraint_validator, alias: tinymce_max_length_validator }
+
+   validator.unique.multiple_emails_with_separator_validator:
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Resources/config/services/bundle/form/form_validator.yml:13]: 3 spaces?

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
I think it's time to squash your commit and remove merge commits.

File: tests/Unit/PrestaShopBundle/Form/Validator/MultipleEmailsWithSeparatorValidatorTest.php [hunk 25]
```diff
@@ -0,0 +1,167 @@
...
+            [MultipleEmailsWithSeparator::INVALID_EMAILS_ERROR_CODE, 'a@@test.com', ';'],
+            [MultipleEmailsWithSeparator::INVALID_EMAILS_ERROR_CODE, 'a@test.com;@test.com', ';'],
+            [MultipleEmailsWithSeparator::INVALID_EMAILS_ERROR_CODE, 'a@test.com;b@test.com;.@@test.com', ';'],
+            [MultipleEmailsWithSeparator::INVALID_EMAILS_ERROR_CODE, 'atest.com', ','],
+            [MultipleEmailsWithSeparator::INVALID_EMAILS_ERROR_CODE, 'a@test.com,b@test.', ','],
+            [MultipleEmailsWithSeparator::INVALID_EMAILS_ERROR_CODE, 'a@testcom,b@ç test.com,c@test.com', ','],
+            [MultipleEmailsWithSeparator::INVALID_EMAILS_ERROR_CODE, 'a uk.@test.com', ':'],
+            [MultipleEmailsWithSeparator::INVALID_EMAILS_ERROR_CODE, 'a@test.com:b@test 84.cl', ':'],
+            [MultipleEmailsWithSeparator::INVALID_EMAILS_ERROR_CODE, 'a@test.com:b@test.com:@test.com', ':'],
+            [MultipleEmailsWithSeparator::INVALID_EMAILS_ERROR_CODE, 'a@testcom,b@ç test.com,c@test.com', ','],
```
Comment (sowbiba) [inline tests/Unit/PrestaShopBundle/Form/Validator/MultipleEmailsWithSeparatorValidatorTest.php:164]: I'm thinking about some edge cases
`a@testcom,,b@testcom`
Using `@` as separator

**Review by sowbiba — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Hi @PululuK ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!
